cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i

```diff
@@ -437,49 +484,59 @@ template <typename algorithmFPType, CpuType cpu>
 bool OrderedRespHelper<algorithmFPType, cpu>::findBestSplitCategoricalFeature(const algorithmFPType * featureVal, const IndexType * aIdx, size_t n,
                                                                               size_t nMinSplitPart, const algorithmFPType accuracy,
                                                                               const ImpurityData & curImpurity, TSplitData & split,
-                                                                              double minWeightLeaf) const
+                                                                              const double minWeightLeaf, const algorithmFPType totalWeights) const
 {
     DAAL_ASSERT(n >= 2 * nMinSplitPart);
     ImpurityData left;
     ImpurityData right;
-    bool bFound = false;
     algorithmFPType vBest;
+    bool bFound               = false;
     size_t nDiffFeatureValues = 0;
+    auto aResponse            = this->_aResponse.get();
+    auto aWeights             = this->_aWeights.get();
+
     for (size_t i = 0; i < n - nMinSplitPart;)
     {
         ++nDiffFeatureValues;
-        size_t count                = 1;
-        const algorithmFPType first = featureVal[i];
-        const size_t iStart         = i;
-        for (++i; (i < n) && (featureVal[i] == first); ++count, ++i)
-            ;
-        if ((count < nMinSplitPart) || ((n - count) < nMinSplitPart) || (count < minWeightLeaf) || ((n - count) < minWeightLeaf)) continue;
+        size_t count                   = 1;
+        const algorithmFPType firstVal = featureVal[i];
+        const size_t iStart            = i;
+        algorithmFPType leftWeights    = aWeights[aIdx[i]].val;
+        for (++i; (i < n) && (featureVal[i] == firstVal); ++count, ++i)
+        {
+            leftWeights += aWeights[aIdx[i]].val;
+        }
+        if ((count < nMinSplitPart) || ((n - count) < nMinSplitPart) || (leftWeights < minWeightLeaf)
+            || ((totalWeights - leftWeights) < minWeightLeaf))
+            continue;
 
         if ((i == n) && (nDiffFeatureValues == 2) && bFound) break; //only 2 feature values, one possible split, already found
 
-        calcImpurity(aIdx + iStart, count, left);
-        subtractImpurity<algorithmFPType, cpu>(curImpurity.var, curImpurity.mean, left.var, left.mean, count, right.var, right.mean, n - count);
-#ifdef DEBUG_CHECK_IMPURITY
-        if (iStart == 0) checkImpurityInternal(aIdx + count, n - count, right);
-#endif
-        const algorithmFPType v = algorithmFPType(count) * left.var + algorithmFPType(n - count) * right.var;
+        algorithmFPType weights = 0.;
+        calcImpurity(aIdx + iStart, count, left, weights);
+        DAAL_ASSERT(fabs(weights - leftWeights) < 0.001);
```

Comment:
Unclear const. How it was chosen?

Reply:
I suggest to fix it later, if it's not a good decision

Reply:
Ok, but please investigate it later

Reply:
agree